purchases/src/test/java/com/revenuecat/purchases/google/usecase/QueryPurchaseHistoryUseCaseTest.kt

```diff
@@ -0,0 +1,697 @@
+                receivedError = error
+            },
+            withConnectedClient = {
+                timesRetried++
+                it.invoke(mockClient)
+            },
+            executeRequestOnUIThread = {
+                queryPurchaseHistoryStubbing answers {
+                    slot.captured.onPurchaseHistoryResponse(
+                        BillingClient.BillingResponseCode.ERROR.buildResult(),
+                        emptyList()
+                    )
+                }
+
+                it.invoke(null)
+            },
+        )
+
+        useCase.run()
+
+        assertThat(timesRetried).isEqualTo(4) // First attempt plus 3 retries
+        assertThat(receivedError).isNotNull
+        assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.StoreProblemError)
+    }
+
+    @Test
+    fun `If service returns ITEM_UNAVAILABLE, doesn't retry`() {
+        val slot = slot<PurchaseHistoryResponseListener>()
+        val queryPurchaseHistoryStubbing = every {
+            mockClient.queryPurchaseHistoryAsync(
+                any<QueryPurchaseHistoryParams>(),
+                capture(slot),
+            )
+        }
+        var receivedError: PurchasesError? = null
+        var timesRetried = 0
+        val useCase = QueryPurchaseHistoryUseCase(
+            QueryPurchaseHistoryUseCaseParams(
+                mockDateProvider,
+                mockDiagnosticsTracker,
+                BillingClient.ProductType.SUBS,
+            ),
+            { _ ->
+                fail("shouldn't be success")
+            },
+            { error ->
+                receivedError = error
+            },
+            withConnectedClient = {
+                timesRetried++
+                it.invoke(mockClient)
+            },
+            executeRequestOnUIThread = {
+                queryPurchaseHistoryStubbing answers {
+                    slot.captured.onPurchaseHistoryResponse(
+                        BillingClient.BillingResponseCode.ITEM_UNAVAILABLE.buildResult(),
+                        emptyList()
+                    )
+                }
+
+                it.invoke(null)
+            },
+        )
+
+        useCase.run()
+
+        assertThat(timesRetried).isEqualTo(1)
+        assertThat(receivedError).isNotNull
+        assertThat(receivedError!!.code).isEqualTo(PurchasesErrorCode.ProductNotAvailableForPurchaseError)
+    }
+
+    // endregion
+
+    private fun Int.buildResult(): BillingResult {
+        return BillingResult.newBuilder().setResponseCode(this).build()
+    }
+
+    private fun mockConsumeAsync(billingResult: BillingResult) {
+        every {
+            mockClient.consumeAsync(capture(capturedConsumeParams), capture(capturedConsumeResponseListener))
+        } answers {
+            capturedConsumeResponseListener.captured.onConsumeResponse(
+                billingResult,
+                capturedConsumeParams.captured.purchaseToken
+            )
+        }
+    }
+
+    private fun mockRunnables() {
```

Comment:
Many of these could be moved to that base class I suggested. Also, we might want to use something like the `MockHandlerFactory` I created